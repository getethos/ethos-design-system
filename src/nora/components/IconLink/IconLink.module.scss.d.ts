// This file is automatically generated.
// Please do not change this file!
interface CssExports {
  'Active': string;
  'Completed': string;
  'Error': string;
  'Icon': string;
  'IconLinkContainer': string;
  'IconLinkContainerRight': string;
  'LinkAdjusted': string;
  'LinkLeft': string;
  'LinkRight': string;
  'Waived': string;
}
export const cssExports: CssExports;
export default cssExports;
