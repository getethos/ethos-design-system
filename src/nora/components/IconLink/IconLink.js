import PropTypes from 'prop-types'
import React from 'react'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'

// TODO -- we should probably create a generic Link component :)
const Link = ({
  className,
  href,
  onClick,
  children,
  target = '_self',
  ...rest
}) => {
  const handleClick = (ev) => {
    ev.preventDefault()
    if (!!onClick) {
      onClick(ev)
    }
  }
  return (
    <a
      href={href}
      className={className}
      onClick={handleClick}
      target={target}
      data-link-id={rest['data-link-id']}
    >
      {children}
    </a>
  )
}

Link.propTypes = {
  className: PropTypes.string,
  href: PropTypes.string,
  onClick: PropTypes.func,
  children: PropTypes.node,
  target: PropTypes.string,
  'data-link-id': PropTypes.string,
}

export const IconLink = ({
  iconPrefix,
  iconName,
  iconClassName,
  iconContainerClassName,
  textClassName,
  textPosition,
  copy,
  onClick,
  ...rest
}) => {
  if (textPosition === 'left') {
    return (
      <Link
        href="#"
        className={iconContainerClassName}
        onClick={onClick}
        data-link-id={rest['data-link-id']}
      >
        <span className={textClassName}>{copy}</span>
        <FontAwesomeIcon
          className={iconClassName}
          icon={[iconPrefix, iconName]}
        />
      </Link>
    )
  } else {
    return (
      <Link
        href="#"
        className={iconContainerClassName}
        onClick={onClick}
        data-link-id={rest['data-link-id']}
      >
        <FontAwesomeIcon
          className={iconClassName}
          icon={[iconPrefix, iconName]}
        />
        <span className={textClassName}>{copy}</span>
      </Link>
    )
  }
}

IconLink.propTypes = {
  props: PropTypes.shape({
    iconPrefix: PropTypes.string.isRequired,
    iconName: PropTypes.string.isRequired,
    iconClassName: PropTypes.string.isRequired,
    iconContainerClassName: PropTypes.string.isRequired,
    textClassName: PropTypes.string.isRequired,
    textPosition: PropTypes.oneOf(['left', 'right']).isRequired,
    copy: PropTypes.string.isRequired,
    onClick: PropTypes.func,
    'data-link-id': PropTypes.string,
  }),
}
