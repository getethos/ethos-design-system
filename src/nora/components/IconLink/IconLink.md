Icon links are used in Nora on the left and right side navigation. Some do not change and
are always gray (e.g. Decisions, Overview, and Notes). Other icons can represent dynmamic
states such as:

- Completed
- Error
- Waived
- Review
- Required
- Ordered

_In these cases, the icons reflect that state the task, decision, etc., is currently in._

As such, you can declaratively utilize the CSS Modules available in: `IconLink.module.scss`
to reflect the current state, by passing in props from your parent component as that's where
you're state will of coure live.

---

Decision

```jsx
// These styles are fairly generic to the icon list
import styles from './IconLink.module.scss'

// These styles are specific to the Nora application
// e.g. Labs icon is --NoraOrange-400
import noraStyles from './NoraIconLink.module.scss'

const iconContainerClasses = `${noraStyles.LeftNavIconLinkContainer} ${styles.IconLinkContainer}`

;<IconLink
  iconPrefix="far"
  iconName="exchange"
  iconClassName={`${styles.Icon}`}
  iconContainerClassName={iconContainerClasses}
  textClassName={styles.LinkRight}
  textPosition="right"
  copy="Decision"
  onClick={(ev) => {
    console.log('Decision onclick called...')
  }}
/>
```

Overview

```jsx
// These styles are fairly generic to the icon list
import styles from './IconLink.module.scss'

// These styles are specific to the Nora application
// e.g. Labs icon is --NoraOrange-400
import noraStyles from './NoraIconLink.module.scss'

const iconContainerClasses = `${noraStyles.LeftNavIconLinkContainer} ${styles.IconLinkContainer}`

;<IconLink
  iconPrefix="fal"
  iconName="user-check"
  iconClassName={`${styles.Icon}`}
  iconContainerClassName={iconContainerClasses}
  textClassName={styles.LinkRight}
  textPosition="right"
  copy="Overview"
  onClick={(ev) => {
    console.log('Overview onclick called...')
  }}
/>
```

Overview/Active

```jsx
// These styles are fairly generic to the icon list
import styles from './IconLink.module.scss'

// These styles are specific to the Nora application
// e.g. Labs icon is --NoraOrange-400
import noraStyles from './NoraIconLink.module.scss'

const iconContainerClasses = `${noraStyles.LeftNavIconLinkContainer} ${styles.IconLinkContainer}`

;<IconLink
  iconPrefix="fal"
  iconName="user-check"
  iconClassName={`${styles.Icon} ${styles.Active}`}
  iconContainerClassName={iconContainerClasses}
  textClassName={`${styles.LinkRight} ${styles.Active}`}
  textPosition="right"
  copy="Overview"
  onClick={(ev) => {
    console.log('Overview onclick called...')
  }}
/>
```

Notes

```jsx
import styles from './IconLink.module.scss'
import noraStyles from './NoraIconLink.module.scss'
const iconContainerClasses = `${noraStyles.LeftNavIconLinkContainer} ${styles.IconLinkContainer}`

;<IconLink
  iconPrefix="fas"
  iconName="align-left"
  iconClassName={`${styles.Icon}`}
  iconContainerClassName={iconContainerClasses}
  textClassName={styles.LinkRight}
  textPosition="right"
  copy="Notes"
  onClick={(ev) => {
    console.log('Notes onclick called...')
  }}
/>
```

For Review

```jsx
import styles from './IconLink.module.scss'
import noraStyles from './NoraIconLink.module.scss'
const iconContainerClasses = `${noraStyles.LeftNavIconLinkContainer} ${styles.IconLinkContainer}`

;<IconLink
  iconPrefix="far"
  iconName="check-square"
  iconClassName={`${styles.Icon}`}
  iconContainerClassName={iconContainerClasses}
  textClassName={styles.LinkRight}
  textPosition="right"
  copy="For Review"
  onClick={(ev) => {
    console.log('For Review onclick called...')
  }}
/>
```

Active—The text is black and bolded, and the icon continues to represent whatever state we're in.
_So, in the following example, the MVR section is active and it is currently in Required state._

```jsx
import styles from './IconLink.module.scss'
import noraStyles from './NoraIconLink.module.scss'

// Active state
const iconContainerClasses = `${noraStyles.LeftNavIconLinkContainer} ${styles.IconLinkContainer} ${styles.Active}`

;<IconLink
  iconPrefix="fal"
  iconName="asterisk"
  // Active
  iconClassName={`${styles.Icon} ${styles.Active}`}
  iconContainerClassName={iconContainerClasses}
  textClassName={styles.LinkRight}
  textPosition="right"
  copy="MVR"
  onClick={(ev) => {
    console.log('mvr (Active) onclick called...')
  }}
/>
```

Required

```jsx
import styles from './IconLink.module.scss'
import noraStyles from './NoraIconLink.module.scss'

// Required state
const iconContainerClasses = `${noraStyles.LeftNavIconLinkContainer} ${styles.IconLinkContainer}`

;<IconLink
  iconPrefix="fal"
  iconName="asterisk"
  iconClassName={`${styles.Icon}`}
  iconContainerClassName={iconContainerClasses}
  textClassName={styles.LinkRight}
  textPosition="right"
  copy="Required"
  onClick={(ev) => {
    console.log('(Required) onclick called...')
  }}
/>
```

Ordered

```jsx
import styles from './IconLink.module.scss'
import noraStyles from './NoraIconLink.module.scss'

const iconContainerClasses = `${noraStyles.LeftNavIconLinkContainer} ${styles.IconLinkContainer}`

;<IconLink
  iconPrefix="fal"
  iconName="clock"
  iconClassName={`${styles.Icon}`}
  iconContainerClassName={iconContainerClasses}
  textClassName={styles.LinkRight}
  textPosition="right"
  copy="MVR"
  onClick={(ev) => {
    console.log('mvr (Ordered) onclick called...')
  }}
/>
```

Completed—click to toggle between active and normal

```jsx
import styles from './IconLink.module.scss'
/**
 * You can simulate this by importing these convenience sass classes
 * and setting the `styles.Active`
 * @import '~ethos-design-system/src/nora/components/IconLink/IconLink.module.scss';
 * */
import noraStyles from './NoraIconLink.module.scss'

const YourComponent = () => {
  const [isActive, setIsActive] = React.useState(false)
  const sillyToggle = (ev) => {
    setIsActive(!isActive)
  }

  const iconContainerClasses = `${noraStyles.LeftNavIconLinkContainer} ${styles.IconLinkContainer}`

  const linkClass = `${styles.LinkRight} ${isActive ? styles.Active : ''}`

  const completedIconClass = `${styles.Icon} ${styles.Completed} ${
    isActive ? styles.Active : ''
  }`

  return (
    <IconLink
      iconPrefix="fas"
      iconName="check-square"
      iconClassName={completedIconClass}
      iconContainerClassName={iconContainerClasses}
      textClassName={linkClass}
      textPosition="right"
      copy="Completed"
      onClick={sillyToggle}
    />
  )
}
;<YourComponent />
```

Error—click to toggle between active and normal

```jsx
import styles from './IconLink.module.scss'
/**
 * You can simulate this by importing these convenience sass classes
 * and setting the `styles.Active`
 * @import '~ethos-design-system/src/nora/components/IconLink/IconLink.module.scss';
 * */
import noraStyles from './NoraIconLink.module.scss'

const YourComponent = () => {
  const [isActive, setIsActive] = React.useState(false)
  const sillyToggle = (ev) => {
    setIsActive(!isActive)
  }

  const iconContainerClasses = `${noraStyles.LeftNavIconLinkContainer} ${styles.IconLinkContainer}`

  const linkClass = `${styles.LinkRight} ${isActive ? styles.Active : ''}`

  const errorIconClass = `${styles.Icon} ${styles.Error} ${
    isActive ? styles.Active : ''
  }`

  return (
    <IconLink
      iconPrefix="fal"
      iconName="exclamation-triangle"
      // Error
      iconClassName={errorIconClass}
      iconContainerClassName={iconContainerClasses}
      textClassName={linkClass}
      textPosition="right"
      copy="MVR"
      onClick={sillyToggle}
    />
  )
}
;<YourComponent />
```

Waived—click to toggle between active and normal

```jsx
import styles from './IconLink.module.scss'
/**
 * You can simulate this by importing these convenience sass classes
 * and setting the `styles.Active`
 * @import '~ethos-design-system/src/nora/components/IconLink/IconLink.module.scss';
 * */
import noraStyles from './NoraIconLink.module.scss'

const YourComponent = () => {
  const [isActive, setIsActive] = React.useState(false)
  const sillyToggle = (ev) => {
    setIsActive(!isActive)
  }

  const iconContainerClasses = `${noraStyles.LeftNavIconLinkContainer} ${styles.IconLinkContainer}`

  const linkClass = `${styles.LinkRight} ${isActive ? styles.Active : ''}`

  const waivedIconClass = `${styles.Icon} ${styles.Waived} ${
    isActive ? styles.Active : ''
  }`

  return (
    <IconLink
      iconPrefix="fal"
      iconName="window-close"
      iconClassName={waivedIconClass}
      iconContainerClassName={iconContainerClasses}
      textClassName={linkClass}
      textPosition="right"
      copy="MVR"
      onClick={sillyToggle}
    />
  )
}
;<YourComponent />
```

## Nora—Right Sidebar

Order more evidences

```jsx
import styles from './IconLink.module.scss'
import noraStyles from './NoraIconLink.module.scss'
const iconContainerClasses = `${styles.IconLinkContainer} ${noraStyles.RightNavIconLinkContainer}`

;<IconLink
  iconPrefix="fal"
  iconName="layer-plus"
  iconClassName={`${styles.Icon}`}
  iconContainerClassName={iconContainerClasses}
  textClassName={styles.LinkRight}
  textPosition="right"
  copy="Order more evidences"
  onClick={(ev) => {
    console.log('Order more evidences onclick called...')
  }}
/>
```

Open original PDFs

```jsx
import styles from './IconLink.module.scss'
import noraStyles from './NoraIconLink.module.scss'
const iconContainerClasses = `${styles.IconLinkContainer} ${noraStyles.RightNavIconLinkContainer}`

;<IconLink
  iconPrefix="far"
  iconName="external-link"
  iconClassName={`${styles.Icon}`}
  iconContainerClassName={iconContainerClasses}
  textClassName={styles.LinkRight}
  textPosition="right"
  copy="Open original PDFs"
  onClick={(ev) => {
    console.log('Open original PDFs onclick called...')
  }}
/>
```

Approve application

```jsx
import styles from './IconLink.module.scss'
import noraStyles from './NoraIconLink.module.scss'
const iconContainerClasses = `${styles.IconLinkContainer} ${noraStyles.RightNavIconLinkContainer}`

;<IconLink
  iconPrefix="fal"
  iconName="check"
  iconClassName={`${styles.Icon}`}
  iconContainerClassName={iconContainerClasses}
  textClassName={styles.LinkRight}
  textPosition="right"
  copy="Approve application"
  onClick={(ev) => {
    console.log('Approve application onclick called...')
  }}
/>
```

Decline application

```jsx
import styles from './IconLink.module.scss'
import noraStyles from './NoraIconLink.module.scss'
const iconContainerClasses = `${styles.IconLinkContainer} ${noraStyles.RightNavIconLinkContainer}`

;<IconLink
  iconPrefix="fal"
  iconName="times"
  iconClassName={`${styles.Icon}`}
  iconContainerClassName={iconContainerClasses}
  textClassName={styles.LinkRight}
  textPosition="right"
  copy="Decline application"
  onClick={(ev) => {
    console.log('Decline application onclick called...')
  }}
/>
```

Withdraw

```jsx
import styles from './IconLink.module.scss'
import noraStyles from './NoraIconLink.module.scss'
const iconContainerClasses = `${styles.IconLinkContainer} ${noraStyles.RightNavIconLinkContainer}`

;<IconLink
  iconPrefix="fal"
  iconName="octagon"
  iconClassName={`${styles.Icon}`}
  iconContainerClassName={iconContainerClasses}
  textClassName={styles.LinkRight}
  textPosition="right"
  copy="Withdraw"
  onClick={(ev) => {
    console.log('Withdraw onclick called...')
  }}
/>
```

Back Button Chevron

```jsx
import styles from './IconLink.module.scss'
;<IconLink
  iconPrefix="far"
  iconName="chevron-left"
  iconClassName={`${styles.Icon}`}
  iconContainerClassName={styles.IconLinkContainer}
  textClassName={styles.LinkRight}
  textPosition="right"
  copy="Back"
  onClick={(ev) => {
    console.log('Back button -- onclick called...')
  }}
/>
```

## Right Aligned

Additionally, you can position the icon to the right and have the icon link
flex-end aligned by declaratively using different CSS Module classes:

```jsx
import styles from './IconLink.module.scss'
;<IconLink
  iconPrefix="far"
  iconName="search"
  iconClassName={`${styles.Icon}`}
  iconContainerClassName={styles.IconLinkContainerRight}
  textClassName={styles.LinkLeft}
  textPosition="left"
  copy="Find it…"
  onClick={(ev) => {
    console.log('mvr (Waived) onclick called...')
  }}
/>
```
