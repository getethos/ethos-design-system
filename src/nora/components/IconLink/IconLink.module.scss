.Icon {
  flex: 0 0 auto;
  height: var(--Space-14);
}

.IconLinkContainer {
  display: flex;
  align-items: center;
  justify-content: flex-start;
  height: 20px;
  margin-bottom: var(--Space-4);
  font-size: calc(var(--compact-font-size) * 1px);
  color: var(--NoraNeutral-400);
  height: 32px;
  border-radius: 3px;
  padding-left: var(--Space-14);

  &:hover {
    background: var(--NoraNeutral-100);
    cursor: pointer;
  }
}

.Active {
  color: var(--NoraNeutral-900);
  font-weight: 500;
}

.Error {
  color: var(--NoraRed-400);
}

.Waived {
  color: var(--NoraOrange-400);
}

.Completed {
  color: var(--NoraGreen-400);
}

.IconLinkContainerRight {
  composes: IconLinkContainer;
  justify-content: flex-end;
}

.LinkAdjusted {
  /* Hack: Our current font is vertically aligned too far to the top of the line box */
  margin-bottom: -2px;
  color: inherit;
}

.LinkLeft {
  composes: LinkAdjusted;
  margin-right: var(--Space-10);
}

.LinkRight {
  composes: LinkAdjusted;
  margin-left: var(--Space-10);
}

.Icon {
  color: inherit;
}
