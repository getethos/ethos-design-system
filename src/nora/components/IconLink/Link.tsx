import * as React from 'react'
type LinkProps = {
  className?: string
  href?: string
  onClick?: (...args: any[]) => any
  target?: string
  'data-link-id'?: string
}
export const Link: React.FC<LinkProps> = ({
  className,
  href,
  onClick,
  children,
  target = '_self',
  ...rest
}) => {
  const handleClick = (ev) => {
    ev.preventDefault()
    // eslint-disable-next-line no-extra-boolean-cast
    if (!!onClick) {
      onClick(ev)
    }
  }
  return (
    <a
      href={href}
      className={className}
      onClick={handleClick}
      target={target}
      data-link-id={rest['data-link-id']}
    >
      {children}
    </a>
  )
}
