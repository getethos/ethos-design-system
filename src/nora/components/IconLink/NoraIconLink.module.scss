.LeftNavIconLinkContainer {
  width: 180px;
}
