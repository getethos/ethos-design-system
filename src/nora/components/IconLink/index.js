export { IconLink } from './IconLink.js'
