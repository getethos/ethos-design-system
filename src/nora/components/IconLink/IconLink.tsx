import { IconName, IconPrefix } from '@fortawesome/fontawesome-svg-core'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import * as React from 'react'
import { Link } from './Link'
type IconLinkProps = {
  iconPrefix: string
  iconName: string
  iconClassName: string
  iconContainerClassName: string
  textClassName: string
  textPosition: 'left' | 'right'
  copy: string
  onClick?: (...args: any[]) => any
  'data-link-id'?: string
}
export const IconLink: React.FC<IconLinkProps> = ({
  iconPrefix,
  iconName,
  iconClassName,
  iconContainerClassName,
  textClassName,
  textPosition,
  copy,
  onClick,
  ...rest
}) => {
  if (textPosition === 'left') {
    return (
      <Link
        href="#"
        className={iconContainerClassName}
        onClick={onClick}
        data-link-id={rest['data-link-id']}
      >
        <span className={textClassName}>{copy}</span>
        <FontAwesomeIcon
          className={iconClassName}
          icon={[iconPrefix as IconPrefix, iconName as IconName]}
        />
      </Link>
    )
  } else {
    return (
      <Link
        href="#"
        className={iconContainerClassName}
        onClick={onClick}
        data-link-id={rest['data-link-id']}
      >
        <FontAwesomeIcon
          className={iconClassName}
          icon={[iconPrefix as IconPrefix, iconName as IconName]}
        />
        <span className={textClassName}>{copy}</span>
      </Link>
    )
  }
}
