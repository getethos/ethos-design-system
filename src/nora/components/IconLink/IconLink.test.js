import React from 'react'
import renderer from 'react-test-renderer'
import { IconLink } from './IconLink'

describe('IconLink', () => {
  it('renders default IconLink', () => {
    const tree = renderer
      .create(
        <IconLink
          iconPrefix="fas"
          iconName="check-square"
          iconClassName="iconClassName"
          iconContainerClassName="iconContainerClasses"
          textClassName="linkClassName"
          textPosition="right"
          copy="Completed"
        />
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
