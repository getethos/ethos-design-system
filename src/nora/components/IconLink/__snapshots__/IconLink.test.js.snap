// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconLink renders default IconLink 1`] = `
<a
  className="iconContainerClasses"
  href="#"
  onClick={[Function]}
  target="_self"
>
  <span
    className="linkClassName"
  >
    Completed
  </span>
</a>
`;
